sycl/source/detail/persistent_device_code_cache.cpp

```diff
@@ -128,6 +124,13 @@ void PersistentDeviceCodeCache::putItemToDisc(
                                             sizeof(char *) * Pointers.size(),
                                             Pointers.data(), nullptr);
 
+  size_t i = 0;
+  std::string FileName;
+  do {
+    FileName = DirName + "/" + std::to_string(i++);
+  } while (OSUtil::isPathPresent(FileName + ".bin") ||
+           OSUtil::isPathPresent(FileName + ".lock"));
```

Comment:
I think that race condition is still possible even with the fix:
- thread 1 checks that `0.lock` doesn't exist and it is the case
- thread 2 makes the same check and it is still the case
- thread 2 acquires the lock by creating `0.lock`
- thread 1 fails to acquire the lock

The more I read the code, the less I understand why do we have this `i` counter at all?

Comment below on line 155 says:

> Program binaries built for one or more devices are read from persistent
> cache and returned in form of vector of programs. Each binary program is
> stored in vector of chars.

But device image hash is part of a path, i.e. cache key already:

https://github.com/intel/llvm/blob/5ff82eae8677ca2fc836550308e546883f1aed09/sycl/source/detail/persistent_device_code_cache.cpp#L356-L359

Which means that if two threads got the same `DirName` then they are caching the same program, or a hash collision happened. `i` was probably designed as a method to deal with those collisions ([yes, it was](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/KernelProgramCache.md#persistent-cache-storage-structure)), but it seems to be completely ignored:

https://github.com/intel/llvm/blob/bca61db536e187c9b7b6fdee5ad1509586f8e850/sycl/source/detail/program_manager/program_manager.cpp#L492-L503

Note that we only ever use the very first element, ignoring all other and we do not check if that first element is what we wanted to see.

Reply:
I find it very hard to believe it would be a true hash collision, odds are too low. It's more likely just the same program being written out.  Which is why the <n> is otherwise ignored.

I'll investigate using a mutex instead of a .lock file. That confused me as well. 

In the tests that we have now, it is clear that "cache miss" is just something that is expected.  But as you've seen discussed in the report,  the consequences might be too high. I think we need to try to eliminate them.

Reply:
> I'll investigate using a mutex instead of a .lock file

Lock file still makes sense in case of multiple processes (not that I imagine it to be a widespread scenario). Perhaps we could put "acquire lock file" into a loop and simply wait until we can acquire the lock. But we need to make sure that we can distinguish between a situation when lock file already exists and we couldn't create it for any other reason (filesystem fault or something) to avoid infinite loop at that stage.

Reply:
> I find it very hard to believe it would be a true hash collision, odds are too low.

I also don't believe that report we have is caused by a collision, but it makes me wonder why don't we see cache file created? Even if thread A has acquired the lock and thread B failed, shouldn't have cache file been created by the thread A in that case?

This second iteration of looking into this problem left me all confused once again :)

Reply:
@AlexeySachkov I was able to track it down and it had nothing to do with the lock file. It was simply unable to create the  directory. I switched over to using `std::filesystem` and `make_preferred()` and that seems to have fixed it.  I'm keeping the trace report improvements.